sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveRowLevelCommandAssignments.scala

```diff
@@ -0,0 +1,103 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Cast}
+import org.apache.spark.sql.catalyst.expressions.objects.AssertNotNull
+import org.apache.spark.sql.catalyst.plans.logical.{Assignment, LogicalPlan, MergeIntoTable, UpdateTable}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.COMMAND
+import org.apache.spark.sql.catalyst.util.CharVarcharUtils
+import org.apache.spark.sql.connector.catalog.SupportsRowLevelOperations
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+import org.apache.spark.sql.internal.SQLConf.StoreAssignmentPolicy
+
+/**
+ * A rule that resolves assignments in row-level commands.
+ *
+ * Note that this rule must be run before rewriting row-level commands into executable plans.
+ * This rule does not apply to tables that accept any schema. Such tables must inject their own
+ * rules to resolve assignments.
+ */
+object ResolveRowLevelCommandAssignments extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(COMMAND), ruleId) {
+    case u: UpdateTable if !u.skipSchemaResolution && u.resolved &&
+        supportsRowLevelOperations(u.table) && !u.aligned =>
+      validateStoreAssignmentPolicy()
+      val newTable = u.table.transform {
+        case r: DataSourceV2Relation =>
+          r.copy(output = r.output.map(CharVarcharUtils.cleanAttrMetadata))
+      }
+      val newAssignments = AssignmentUtils.alignAssignments(u.table.output, u.assignments)
+      u.copy(table = newTable, assignments = newAssignments)
+
+    case u: UpdateTable if !u.skipSchemaResolution && u.resolved && !u.aligned =>
+      resolveAssignments(u)
+
+    case m: MergeIntoTable if !m.skipSchemaResolution && m.resolved =>
+      resolveAssignments(m)
+  }
+
+  private def validateStoreAssignmentPolicy(): Unit = {
+    // SPARK-28730: LEGACY store assignment policy is disallowed in data source v2
+    if (conf.storeAssignmentPolicy == StoreAssignmentPolicy.LEGACY) {
+      throw QueryCompilationErrors.legacyStoreAssignmentPolicyError()
+    }
+  }
+
+  private def supportsRowLevelOperations(table: LogicalPlan): Boolean = {
+    EliminateSubqueryAliases(table) match {
+      case DataSourceV2Relation(_: SupportsRowLevelOperations, _, _, _, _) => true
+      case _ => false
+    }
+  }
+
+  private def resolveAssignments(p: LogicalPlan): LogicalPlan = {
```

Comment:
Copied from `ResolveOutputRelation` to preserve the existing behavior for data sources that rely on custom implementation.